src/server/game/Entities/AreaTrigger/AreaTrigger.cpp

```diff
@@ -100,14 +148,258 @@ void AreaTrigger::Update(uint32 p_time)
         Remove(); // expired
 
     WorldObject::Update(p_time);
+    _timeSinceCreated += p_time;
+
+    sScriptMgr->OnAreaTriggerEntityUpdate(this, p_time);
+
+    switch (GetTemplate()->GetType())
+    {
+        case AREATRIGGER_TYPE_SPHERE:
+            SearchUnitInSphere();
+            break;
+        case AREATRIGGER_TYPE_BOX:
+            SearchUnitInBox();
+            break;
+        case AREATRIGGER_TYPE_POLYGON:
+            SearchUnitInPolygon();
+            break;
+        case AREATRIGGER_TYPE_CYLINDER:
+            SearchUnitInCylinder();
+            break;
+        default:
+            break;
+    }
 }
 
 void AreaTrigger::Remove()
 {
     if (IsInWorld())
     {
+        if (Unit* caster = GetCaster())
+            caster->_UnregisterAreaTrigger(this);
+
+        for (ObjectGuid insideUnitGuid : _insideUnits)
+            if (Unit* leavingUnit = ObjectAccessor::GetUnit(*this, insideUnitGuid))
+                sScriptMgr->OnAreaTriggerEntityUnitExit(this, leavingUnit);
+
+        sScriptMgr->OnAreaTriggerEntityRemove(this);
+
         SendObjectDeSpawnAnim(GetGUID());
         RemoveFromWorld();
         AddObjectToRemoveList();
     }
 }
+
+void AreaTrigger::SearchUnitInSphere()
+{
+    std::list<Unit*> targetList;
+
+    Trinity::AnyUnitInObjectRangeCheck check(this, GetTemplate()->MaxSearchRadius);
```

Comment:
You will need to extend AnyUnitInObjectRangeCheck constructor to add 3rd argument check3D and pass false to make its inner check IsWithinDistInMap also check only 2D distance here (but it must default to true to keep its current behavior for all other uses in TC)